crates/ruma-events/src/room/message.rs

```diff
@@ -915,46 +917,103 @@ pub(crate) fn parse_markdown(text: &str) -> Option<String> {
         return None;
     }
 
+    // If the markdown contains only inline tags, i.e. if there is only one paragraph, remove the
+    // wrapping paragraph, as instructed by the spec.
+    let first_event_is_paragraph_start =
+        parser_events.front().is_some_and(|event| matches!(event, Event::Start(Tag::Paragraph)));
+    let last_event_is_paragraph_end =
+        parser_events.back().is_some_and(|event| matches!(event, Event::End(TagEnd::Paragraph)));
+
+    if first_event_is_paragraph_start && last_event_is_paragraph_end {
+        found_first_paragraph = false;
+        let has_several_blocks = parser_events.iter().any(|ref event| {
+            if matches!(event, Event::Start(Tag::Paragraph)) {
+                if found_first_paragraph {
+                    true
+                } else {
+                    found_first_paragraph = true;
+                    false
+                }
+            } else if let Event::Start(tag) = event {
+                is_block_tag(tag)
+            } else {
+                false
+            }
+        });
+
+        if !has_several_blocks {
+            parser_events.pop_front();
```

Comment:
Speaking of iterators, I was wondering if instead of removing the events from the `VecDeque`, it would be more performant to just skip them in the iterator below. Something like:

```rust
let mut skip = 0;
let mut take = parser_events.len();

if first_event_is_paragraph_start && last_event_is_paragraph_end {
  // code…

  if !has_several_blocks {
    skip = 1;
    take -= 2;
  }
}

let mut html_body = String::new();
pulldown_cmark::html::push_html(&mut html_body, parser_events.into_iter().skip(skip).take(take));
```

Reply:
Actually, since it's a double-ended iterator, I could do a bit simpler, like:

```rust
let mut iter = parser_events.into_iter();

if !has_several_blocks {
  iter.next();
  iter.next_back();
}
```

Reply:
Sure, if that works.